src/layers/learning/embedding.cpp

```diff
@@ -37,74 +37,26 @@ void embedding_layer<data_layout::DATA_PARALLEL,El::Device::CPU>::setup_matrices
   m_dictionary_gradient = StarMat<El::Device::CPU>(grid);
 }
 
-template <>
-void embedding_layer<data_layout::DATA_PARALLEL,El::Device::CPU>::setup_dims() {
-  Layer::setup_dims();
-
-  // Make sure input dimensions are valid
-  if (this->get_input_size() != 1) {
-    const auto& input_dims = this->get_input_dims();
-    std::ostringstream err;
-    err << get_type() << " layer \"" << get_name() << "\" "
-        << "recieved an input tensor with invalid dimensions "
-        << "(expected 1, got ";
-    for (size_t i = 0; i < input_dims.size(); ++i) {
-      err << (i > 0 ? "x" : "") << input_dims[i];
-    }
-    err << ")";
-    LBANN_ERROR(err.str());
-  }
-
-  // Output is size of embedding vector
-  this->set_output_dims({static_cast<int>(m_embedding_size)});
-
-}
-
-template <>
-void embedding_layer<data_layout::DATA_PARALLEL,El::Device::CPU>::setup_data() {
-  Layer::setup_data();
-
-  // Make sure layer has weights for dictionary
-  if (this->m_weights.size() != 1) {
-    std::ostringstream err;
-    err << "attempted to setup "
-        << this->get_type() << " layer \"" << this->get_name() << "\" "
-        << "with an invalid number of weights "
-        << "(expected 1, "
-        << "found " << this->m_weights.size() << ")";
-    LBANN_ERROR(err.str());
-  }
-
-  // Initialize dictionary
-  auto& dict = *m_weights[0];
-  auto matrix_dist = get_prev_activations().DistData();
-  matrix_dist.colDist = El::STAR;
-  matrix_dist.rowDist = El::STAR;
-  dict.set_dims({static_cast<int>(m_embedding_size)},
-                {static_cast<int>(m_dictionary_size)});
-  dict.set_matrix_distribution(matrix_dist);
-
-  // Initialize gradient w.r.t. dictionary
-  m_dictionary_gradient.Resize(m_embedding_size, m_dictionary_size);
-
-}
-
 template <>
 void embedding_layer<data_layout::DATA_PARALLEL,El::Device::CPU>::fp_compute() {
 
   // Local data
-  const auto& local_dict = m_weights[0]->get_values().LockedMatrix();
-  const auto& local_input = get_local_prev_activations();
-  auto& local_output = get_local_activations();
+  const auto& local_dict = dynamic_cast<const CPUMat&>(m_weights[0]->get_values().LockedMatrix());
+  const auto& local_input = dynamic_cast<const CPUMat&>(get_local_prev_activations());
+  auto& local_output = dynamic_cast<CPUMat&>(get_local_activations());
   const auto& local_width = local_input.Width();
 
   // Populate output matrix with appropriate columns of dictionary
   CPUMat dict_v, output_v;
   for (El::Int col = 0; col < local_width; ++ col) {
-    const El::Int ind = static_cast<El::Int>(local_input(0, col));
-    El::LockedView(dict_v, local_dict, El::ALL, El::IR(ind));
     El::View(output_v, local_output, El::ALL, El::IR(col));
-    El::Copy(dict_v, output_v);
+    const El::Int ind = static_cast<El::Int>(std::floor(local_input(0, col)));
+    if (0 <= ind && ind < static_cast<El::Int>(m_num_embeddings)) {
+      El::LockedView(dict_v, local_dict, El::ALL, El::IR(ind));
+      El::Copy(dict_v, output_v);
```

Comment:
PyTorch's API does not guarantee that the pad index will output zeros, just that the weights are initialized to zero and that the embedding layer does not contribute gradients. If the pad embedding gets gradients from somewhere else, then it can be non-zero.

The LBANN implementation will manually zero out the weights for the pad index, so we match PyTorch's behavior.